### PR TITLE
[InstCombine] Check for undef before transforming lshr (mul (X, 2^N + 1)), N -> add (X, lshr(X, N))

### PR DESCRIPTION
**Pull request by AZero13**
Alive2 didn't catch the partial undef.

https://alive2.llvm.org/ce/z/VkqNSN

### CODE REVIEW

**Comment by AtariDreams** [comment]
@nikic Given how this is a multi-use, it would make sense to put freeze here, no?

**Comment by AtariDreams** [comment]
@dtcxzyw Thoughts?

**Comment by AtariDreams** [comment]
@nikic Is this acceptable now?

**Comment by AtariDreams** [comment]
@dtcxzyw Requesting you review on this...

**Comment by AtariDreams** [comment]
@MaskRay @TNorthover Would any of you please look at this?

**Comment by AtariDreams** [comment]
@dtcxzyw As this is not a new transformation but rather a precaution/bugfix for an existing one, I would like this to be backported to 19.x

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp [hunk 1]
```diff
@@ -1499,7 +1499,7 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
           return replaceInstUsesWith(I, X);
 
         // lshr (mul nuw (X, 2^N + 1)), N -> add nuw (X, lshr(X, N))
-        if (Op0->hasOneUse()) {
+        if (Op0->hasOneUse() && isGuaranteedNotToBeUndef(X, &AC, &I, &DT)) {
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1502]: We should do `if (!isGuaranteedNotToBeUndef) CreateFreeze` instead.

Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1502]: > We should do `if (!isGuaranteedNotToBeUndef) CreateFreeze` instead.

That creates an extra instruction, so we end up with more instructions than we started with.

**Comment by nikic** [comment]
TBH I don't really want to fix partial-undef multi-use miscompiles. We don't have any evidence that they have any read-world relevance.

We typically do not even fix full-undef multi-use miscompiles in pre-existing transforms (we only don't accept them in new ones).

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp (hunk 1 above)
Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1502]: It's ok if the extra instruction is freeze.

**Comment by AtariDreams** [comment]
> TBH I don't really want to fix partial-undef multi-use miscompiles. We don't have any evidence that they have any read-world relevance.
> 
> We typically do not even fix full-undef multi-use miscompiles in pre-existing transforms (we only don't accept them in new ones).

Honestly, that's valid. I will just close this for now and do some more research.

**Comment by dtcxzyw** [comment]
> TBH I don't really want to fix partial-undef multi-use miscompiles. We don't have any evidence that they have any read-world relevance.
> 
> We typically do not even fix full-undef multi-use miscompiles in pre-existing transforms (we only don't accept them in new ones).

@nikic Can you clarify this policy in instcombine contributor guide?

**Comment by AtariDreams** [comment]
The issue I have is that by masking a partial undef, you turn a partial undef into a full undef

**Pull request closed without merging**